nvmeadm/tests/discovery_test.rs

```diff
@@ -1,7 +1,147 @@
+  BlockSize    4096
+[Nvmf]
+  AcceptorPollRate 10000
+  ConnectionScheduler RoundRobin
+[Transport]
+  Type TCP
+  # reduce memory requirements
+  NumSharedBuffers 32
+[Subsystem1]
+  NQN nqn.2019-05.io.openebs:disk2
+  Listen TCP 127.0.0.1:NVMF_PORT
+  AllowAnyHost Yes
+  SN MAYASTOR0000000001
+  MN NEXUSController1
+  MaxNamespaces 1
+  Namespace Malloc0 1
+# although not used we still have to reduce mem requirements for iSCSI
+[iSCSI]
+  MaxSessions 1
+  MaxConnectionsPerSession 1
+";
+
+static CONFIG_FILE: &str = "/tmp/nvmeadm_nvmf_target.config";
+
+static SERVED_DISK_NQN: &str = "nqn.2019-05.io.openebs:disk2";
+
+static TARGET_PORT: u32 = 9523;
+
+// Writes out a config file for spdk, but with the specified port for nvmf to
+// use
+fn create_config_file(config_file: &str, nvmf_port: &str) {
+    let path = Path::new(config_file);
+    let mut config = match File::create(&path) {
+        Err(reason) => panic!(
+            "Unable to create {}: {}",
+            path.display(),
+            reason.to_string()
+        ),
+        Ok(config) => config,
+    };
+
+    let after = CONFIG_TEXT.replace("NVMF_PORT", nvmf_port);
+
+    match config.write_all(after.as_bytes()) {
+        Err(reason) => panic!(
+            "Unable to write to {}: {}",
+            path.display(),
+            reason.to_string()
+        ),
+        Ok(_) => println!("Wrote to file"),
+    }
+}
+
+// Waits for spdk to start up and accept connections on the specified port
+fn wait_for_spdk_ready(listening_port: u32) -> Result<(), String> {
+    let dest = format!("127.0.0.1:{}", listening_port);
+    let socket_addr: SocketAddr = dest.parse().expect("Badly formed address");
+
+    let mut bound = false;
+    for _index in 1 .. 20 {
+        let result = TcpStream::connect_timeout(
+            &socket_addr,
+            Duration::from_millis(100),
+        );
+        bound = result.is_ok();
+        if bound {
+            break;
+        }
+        thread::sleep(Duration::from_millis(101));
+    }
+
+    if bound {
+        Ok(())
+    } else {
+        Err(format!(
+            "spdk listening port ({}) not found",
+            listening_port
+        ))
+    }
+}
+
+pub struct NvmfTarget {
+    /// The std::process::Child for the process running spdk
+    pub spdk_proc: std::process::Child,
+}
+
+impl NvmfTarget {
+    pub fn new(config_file: &str, nvmf_port: &str) -> Result<Self, Error> {
+        create_config_file(config_file, nvmf_port);
+        let spdk_proc = Command::new("../target/debug/spdk")
+            .arg("-c")
+            .arg(CONFIG_FILE)
+            .spawn()
+            .expect("Failed to start spdk!");
+
+        wait_for_spdk_ready(TARGET_PORT).expect("spdk not ready");
+
+        let _ = DiscoveryBuilder::default()
+            .transport("tcp".to_string())
+            .traddr("127.0.0.01".to_string())
+            .trsvcid(TARGET_PORT)
+            .build()
+            .unwrap();
+
+        println!("SPDK started");
```

Comment:
s/println/info/g ?

Reply:
I guess the println! stuff can become info! ...